### PR TITLE
EN-4605 Add support for Shineretrofits custom shipping

### PR DESCRIPTION
**Pull request by dumega**
# Description
Add support for Shineretrofits custom shipping, that's a custom shipping module built by the merchant Shineretrofits

Fixes: https://boltpay.atlassian.net/browse/EN-4605

#changelog Add support for Shineretrofits custom shipping

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Comment by njfaries** [comment]
@dumega is this still a draft PR or is it ready for review?

File: Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        if($this->_appState->getAreaCode() == 'frontend'){
+            return false;
+        }
+        if($this->_appState->getAreaCode() == 'webrest_api'){
+            return false;
+        }
+        if (!$this->isBoltCalculation()) {
+            return false;
+        }
```
Comment (danac-gs) [inline Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php:106]:
```suggestion
        if ($result !== false) {
	    return $result;
	}        
	if (
	    !$subject->isActive()
	    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_FRONTEND
	    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_WEBAPI_REST
	    || !$this->isBoltCalculation()
	) {
	    return false;
	}
```

**Review by danac-gs — CHANGES_REQUESTED** [review]

File: Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+    
+    /**
+     * Checks if we do calculation for Bolt plugin.
+     *
+     * @return bool
+     */
+    protected function isBoltCalculation()
+    {
+        if ($this->_appState->getAreaCode() !== 'webapi_rest') {
```
[the remark this answers is not in the document]
Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: Done

File: Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            return false;
+        }
+        if (!empty($this->_restRequest)) {
```
[the remark this answers is not in the document]
Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: Removed

File: Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * The original method prohibits using Shineretrofits shipping method shipping method for webapi_rest,
+     * so it returns false for Bolt backoffice order.
+     * We need to check if calculation was called for Bolt and return rates if so.
+     */
```
[the remark this answers is not in the document]
Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php:92]: Added

File: Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php (hunk 1 above)
Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: Done, but `webrest_api` does not exist in the M2 core, it seems to be a custom AreaCode created by the dev of this merchant.
So I change the code to

```
if (
    !$subject->isActive()
    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_FRONTEND
    || $this->_appState->getAreaCode() == 'webrest_api'
    || !$this->isBoltCalculation()
) {
    return false;
}
```

Comment (danac-gs) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: @dumega `webapi_rest` exists here
https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/App/Area.php#L23

Are you sure it's not a typo on their end? It's very strange to have a need of crating a custom area code.

Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: @danac-gs In their code, it is **webrest_api**, not **webapi_rest** . And I can ask CSM to confirm with the merchant

Comment (dumega) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: @danac-gs 
Below is the original code

```
if($this->state->getAreaCode() == 'frontend'){
    return false;
}
if($this->state->getAreaCode() == 'webrest_api'){
    return false;
}
if($this->state->getAreaCode() == 'webapi_rest'){
    return false;
}
```

For webapi_rest, I create function isBoltCalculation() to cover the case.
https://github.com/BoltApp/bolt-magento2/pull/1252/files#diff-9440d2cc9ac6b28f4535e7f561ad104c31e427c677cee31d4e248762e4c7622bR64

So it seems that webrest_api is intended. I think we would better follow the same conditions.

Comment (danac-gs) [reply Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php]: I agree. Thanks.

**Review by danac-gs — APPROVED** [review]

**Review by njfaries — APPROVED** [review]

**Pull request merged**